man/netperf-wrapper.1
.TH NETPERF-WRAPPER "1" "November 2014" "Netperf-wrapper v0.9.1-git." "User Commands"
.SH NAME
Netperf-wrapper \- manual page for Netperf-wrapper v0.9.1-git
.SH SYNOPSIS
.B netperf-wrapper
[\fIoptions\fR] \fI<host|test|input file \fR...\fI>\fR
.SH INVOCATION
When run, netperf-wrapper must be supplied either (a) a test name and one or
more host names to connect to, or (b) one or more input files containing data
from previous runs to post-process.
.PP
Test names, hostnames and input file names can all be specified as unqualified
arguments, and netperf-wrapper will do its best to guess which is which. For
each argument, if it is an existing file, it is assumed to be an input file, if
it is the name of an existing test configuration it's assumed to be a test name,
and if neither of those are true, it is assumed to be a host name. The \fB-i\fR
and \fB-H\fR switches can be used to explicitly specify the interpretation of an
argument.

.SH DESCRIPTION
Netperf-wrapper is a wrapper around \fI netperf \fR and similar tools to
run predefined tests and aggregate and plot the results. It defines several
tests that can be run against one or more hosts, primarily targeted at testing
for the presence of bufferbloat under various conditions.
.PP
The aggregated data is saved in (gzipped) JSON format for later processing
and/or import into other tools. The JSON format is documented below.
.PP
Apart from the JSON format, the data can be output as csv values, emacs org
mode tables or plots. Each test can specify several different plots, including
time-series plots of the values against each other, as well as CDF plots of
(e.g.) ping times.
.PP
Plots can be output to the formats supported by matplotlib by specifying the
output filename with \fB-o\fR \fIoutput.{png,ps,pdf,svg}\fR. If no output file
is specified, the plot is diplayed using matplotlib's interactive plot browser,
which also allows saving of the output (in .png format).

.SH REQUIREMENTS
Netperf-wrapper runs on Python, versions 2.6 through 3.3. Plotting requires a
functional matplotlib installation (but everything else can run without
matplotlib). For the interactive plot viewer, a graphical display (and suitably
configured matplotlib) is required.
.PP
Most tests employ the \fInetperf\fR benchmarking tool to run the tests. Version
2.6 or higher is required, and netperf must be compiled with the
\fB\-\-enable\-demo\fR option passed to ./configure. Some tests use \fIiperf\fR
in addition to, or instead of netperf. Both tools must be available in the
\fIPATH\fR.
.PP
For ICMP ping measurements, the version of \fIping\fR employed must support
output timestamping (the \fB\-D\fR parameter to GNU ping). This is \fBnot\fR
supported by the BSD and OSX versions of ping. As an alternative to the regular
ping command, the \fIfping\fR utility (see \fIfping.org\fR) can be employed. In
that case fping must be version 3.5 or greater. Netperf-wrapper will attempt to
detect the presence of fping in the \fIPATH\fR and check for support for the
\fB\-D\fR parameter. If this check is successful, fping will be employed for
ping data, otherwise the system ping will be used.

.SH OPTIONS
.TP
\fB\-o\fR OUTPUT, \fB\-\-output\fR=\fIOUTPUT\fR
File to write processed output to (default standard
out). The JSON data file is written to the same
directory as this file, if provided. Otherwise, the
data file is written to the current directory.
.TP
\fB\-i\fR INPUT, \fB\-\-input\fR=\fIINPUT\fR
File to read input from (instead of running tests).
Input files can also be specified as unqualified
arguments without using the \fB\-i\fR switch.
.TP
\fB\-f\fR FORMAT, \fB\-\-format\fR=\fIFORMAT\fR
Select output format (plot, csv, org_table, stats). Default
is no processed output (just writes the JSON data
file).
.TP
\fB\-p\fR PLOT, \fB\-\-plot\fR=\fIPLOT\fR
Select which plot to output for the given test
(implies \fB\-f\fR plot). Use the \fB\-\-list\-plots\fR option to see
available plots.
.TP
\fB\-t\fR TITLE, \fB\-\-title\-extra\fR=\fITITLE\fR
Text to add to plot title and data file name.
.TP
\fB\-n\fR NOTE, \fB\-\-note\fR=\fINOTE\fR
Add arbitrary text as a note to be stored in the JSON data file
(under the \fINOTE\fR key in the metadata object).
.TP
\fB\-r\fR RCFILE, \fB\-\-rcfile\fR=\fIRCFILE\fR
Load configuration data from \fIRCFILE\fR (default \fI~/.netperf-wrapperrc\fR). See
section below for information on the rc file format.
.TP
\fB\-x\fR, \fB\-\-extended\-metadata\fR
Collect extended metadata and store it with the data file.
May include details of your machine you don't want to distribute; see the
section on the data format below.
.TP
\fB\-\-remote\-metadata\fR=\fIHOSTNAME\fR
Collect extended metadata from a remote host. \fIHOSTNAME\fR is passed verbatim
to ssh, so can include hosts specified in ~/.ssh/config. Note that gathering the
data can take some time, since it involves executing several remote commands.
This option can be specified multiple times and implies \fB--extended-metadata\fR.
.TP
\fB--gui\fR
Run the netperf-wrapper GUI. All other options are used as defaults in the GUI,
but can be changed once it is running. See section below for a description of
the GUI.
.TP
\fB--new-gui-instance\fR
Start a new GUI instance. Otherwise, netperf-wrapper will try to connect to an
already running GUI instance and have that load any new data files specified as
arguments. Implies \fB--gui\fR when passed on the command line, but not when set in
the rc file. Note that when multiple GUI instances are running, there is no
guarantee as to which instance will get a subsequent open request (if run again
without \fB--new-gui-instance\fR).
.TP
\fB--gui-no-defer\fR 
Normally, the GUI defers redrawing plots until they are needed to avoid
redrawing all open plots every time an option changes. This switch turns off
that optimisation in favour of always redrawing everything straight away. This
is useful when loading a bunch of plots from the command line and then wanting
to flip through them without drawing delay.
.TP
\fB\-b\fR BATCH_NAME, \fB\-\-batch\-name\fR=\fIBATCH_NAME\fR
Run test batch BATCH_NAME (must be specified in a batch file loaded by the
\fB--batch-file\fR option). Can be supplied multiple times.
.TP
\fB\-B\fR BATCH_FILE, \fB\-\-batch\-file\fR=\fIBATCH_FILE\fR
Load batch file BATCH_FILE. Can be specified multiple times, in which case the
files will be combined (with identically-named sections being overridden by
later files). See appropriate section below for an explanation of the batch file
format.
.TP
\fB\-\-batch\-override\fR=\fIkey=value\fR
Override parameter 'key' in the batch config and set it to 'value'. The key name
will be case folded to lower case. Can be specified multiple times.
.TP
\fB--batch-dry-run\fR 
Dry batch run. Prints what would be done, but doesn't actually run any tests.
.TP
\fB\-\-batch\-repetitions\fR=\fIREPETITIONS\fR
Shorthand for \fB--batch-override\fR \fI'repetitions=REPETITIONS'\fR.
.TP
\fBTest configuration:\fR
.PP
These options affect the behaviour of the test being run and have no
effect when parsing input files.
.TP
\fB\-H\fR HOST, \fB\-\-host\fR=\fIHOST\fR
Host to connect to for tests. For tests that support
it, multiple hosts can be specified by supplying this
option multiple times. Hosts can also be specified as
unqualified arguments; this parameter guarantees that
the argument be interpreted as a host name (rather
than being subject to auto\-detection between input
files, hostnames and test names).
.TP
\fB\-\-local-bind\fR=\fIIP\fR
Local hostname or IP address to bind to (for test tools that support this).
.TP
\fB\-l\fR LENGTH, \fB\-\-length\fR=\fILENGTH\fR
Base test length (some tests may add some time to
this).
.TP
\fB\-s\fR STEP_SIZE, \fB\-\-step\-size\fR=\fISTEP_SIZE\fR
Measurement data point step size.
.TP
\fB\-d\fR DELAY, \fB\-\-delay\fR=\fIDELAY\fR
Number of seconds to delay parts of test (such as
bandwidth loaders).
.TP
\fB\-4\fR, \fB\-\-ipv4\fR
Use IPv4 for tests (some tests may ignore this).
.TP
\fB\-6\fR, \fB\-\-ipv6\fR
Use IPv6 for tests (some tests may ignore this).
.TP
\fB\-\-socket\-timeout\fR=\fISOCKET_TIMEOUT\fR
Socket timeout (in seconds) used for UDP delay measurement, to prevent
stalls on packet loss. Only enabled if the installed netperf version is
detected to support this (requires SVN version of netperf).
.IP
For the default value, see the output of netperf-wrapper -h. The value of this
parameter is an implicit upper bound on how long a round-trip time that can be
measured. As such you may need to adjust it if you are experiencing latency
above the default value. Set to 0 to disable.
.TP
\fB\-\-test\-parameter\fR=\fIkey=value\fR
Arbitrary test parameter in key=value format. Key will be case folded to lower
case. Some test configurations may alter behaviour based on values passed as
test parameters. Additionally, the values are stored with the results metadata,
and so can be used for arbitrary resultset categorisation. Can be specified
multiple times.
.TP
\fB\-\-swap\-up\-down\fR
Switch upstream and downstream directions for data transfer. This means
that 'upload' will become 'download' and vice versa. Works by exchanging netperf
TCP_MAERTS and TCP_STREAM parameters, so only works for tests that employ these
as their data transfer, and only for the TCP streams.
.TP
\fBPlot configuration:\fR
.PP
These options are used to configure the appearance of plot output and
only make sense combined with \fB\-f\fI plot\fR.
.TP
\fB\-z\fR, \fB\-\-zero\-y\fR
Always start y axis of plot at zero, instead of autoscaling the axis (also disables log scales). Autoscaling is still enabled for the upper bound.
.TP
\fB\-I\fR, \fB\-\-invert\-latency\-y\fR
Invert the y-axis for latency data series (making plots show 'better values upwards').
.TP
\fB\-\-disable\-log\fR
Disable log scales on plots.
.TP
\fB\-\-norm\-factor\fR=\fIFACTOR\fR
Factor to normalise data by. I.e. divide all data points by this value. Can be
specified multiple times, in which case each value corresponds to a data series.
.TP
\fB\-\-scale\-data\fR=\fISCALE_DATA\fR
Additional data files to consider when scaling the
plot axes (for plotting several plots with identical
axes). Note, this displays only the first data set, but with axis
scaling taking into account the additional data sets. Can be supplied
multiple times; see also \fB\-\-scale\-mode\fR.
.TP
\fB\-S\fR, \fB\-\-scale\-mode\fR
Treat file names (except for the first one) passed as
unqualified arguments as if passed as \fB\-\-scale\-data\fR
(default as if passed as \fB\-\-input\fR).
.TP
\fB\-\-concatenate\fR
Concatenate multiple result sets into one data series. This means that each data
file will have its time axis shifted by the preceding series duration and
appended to the first data set specified. Only works for data sets from the same
test, obviously.
.TP
\fB\-\-absolute\-time\fR
Plot data points with absolute UNIX time on the x-axis. This requires the
absolute starting time for the test run to be stored in the data file, and so it
won't work with data files that predates this feature.
.TP
\fB\-\-subplot\-combine\fR
When plotting multiple data series, plot each one on a separate subplot instead
of combining them into one plot. This mode is not supported for all plot types,
and only works when \fB--scale-mode\fR is disabled.
.TP
\fB\-\-no\-print\-n\fR
Do not print the number of data points on combined plots. When using plot types
that combines results from several test runs, the number of data series in each
combined data point is normally added after the series name, (n=X) for X data
series. This option turns that off.
.TP
\fB\-\-no\-annotation\fR
Exclude annotation with hostnames, time and test
length from plots.
.TP
\fB\-\-no\-title\fR
Exclude title from plots.
.TP
\fB\-\-override\-title\fR=\fITITLE\fR 
Override plot title with this string. Completely discards the configured title
(from the test configuration), as well as the title stored in the data set, and
replaces it with the value supplied here. This is useful to override the plot
title \fIat the time of plotting\fR, for instance to add a title to an aggregate
plot from several data series. When this parameter is specified,
\fB--no-title\fR has no effect.
.TP
\fB\-\-no\-markers\fR
Don't use line markers to differentiate data series on plots.
.TP
\fB\-\-no\-legend\fR
Exclude legend from plots.
.TP
\fB\-\-filter\-legend\fR
Filter legend labels by removing the longest common substring from all entries.
This is not particularly smart, so should be used with care.
.TP
\fB\-\-filter\-regexp\fR=\fIREGEXP\fR
Filter the plot legend by the supplied regular expression. Note that for
combining several plot results, the regular expression is also applied before
the grouping logic, meaning that a too wide filter can mess up the grouping.
.TP
\fB\-\-figure\-width\fR=\fIFIG_WIDTH\fR 
Figure width in inches. Used when saving plots to file and for default size of
the interactive plot window.
.TP
\fB\-\-figure\-height\fR=\fIFIG_HEIGHT\fR 
Figure height in inches. Used when saving plots to file and for default size of
the interactive plot window.
.TP
\fB\-\-figure\-dpi\fR=\fIFIG_DPI\fR 
Figure DPI. Used when saving plots to raster format files.
.TP
\fB\-\-no\-matplotlibrc\fR
Don't load included matplotlibrc values. Use this if autodetection of custom
matplotlibrc fails and netperf-wrapper is inadvertently overriding rc values.
.IP
.TP
\fBTest tool-related options:\fR
.TP
\fB\-\-control\-host\fR=\fIHOST\fR
Hostname for the test control connection (for test tools that support this).
Default: First hostname of test target.

When running tests that uses D-ITG as a test tool (such as the voip-* tests),
this switch controls where netperf-wrapper will look for the D-ITG control
server (see section below on running tests with D-ITG). For Netperf-based tests,
this option is passed to Netperf to control where to point the control
connection. This is useful to, for instance, to run the control server
communication over a separate control network so as to not interfere with test
traffic.
.TP
\fB\-\-control\-local-bind\fR=\fIIP\fR
Local hostname or IP to bind control connection to (for test tools that support
it; currently netperf). If not supplied, the value for --local-bind will be
used. Note that if this value is passed but \fB--local-bind\fR is \fBnot\fR,
netperf will use the value specified here to bind the data connections to as
well.
.TP
\fB\-\-netperf\-control\-port\fR=\fIPORT\fR 
Port for Netperf control server. Default: 12865.
.TP
\fB\-\-ditg\-control\-port\fR=\fIPORT\fR 
Port for D-ITG control server. Default: 8000.
.TP
\fB\-\-ditg\-control\-secret\fR=\fISECRET\fR 
Secret for D-ITG control server authentication. Default: ''.
.TP
\fB\-\-http\-getter\-urllist\fR=\fIFILENAME\fR
When running HTTP tests, the http-getter tool is used to fetch URLs (see
https://github.com/tohojo/http-getter). This option specifies the filename
containing the list of HTTP URLs to get. Can also be a URL, which will then be
downloaded as part of each test iteration. If not specified, this is set to
http://<hostname>/filelist.txt where <hostname> is the first test hostname.
.TP
\fB\-\-http\-getter\-dns\-servers\fR=\fIDNS_SERVERS\fR
DNS servers to use for http-getter lookups. Format is
host[:port][,host[:port]]... This option will only work if libcurl supports it
(needs to be built with the ares resolver). Default is none (use the system
resolver).
.TP
\fB\-\-http\-getter\-timeout\fR=\fIMILLISECONDS\fR
Timeout for HTTP connections. Default is to use the test length.
.TP
\fB\-\-http\-getter\-workers\fR=\fINUMBER\fR
Number of workers to use for getting HTTP urls. Default is 4.
.IP
.TP
\fBMisc and debugging options:\fR
.TP
\fB\-L\fR LOG_FILE, \fB\-\-log\-file\fR=\fILOG_FILE\fR
Write debug log (test program output) to log file.
.TP
\fB\-\-list\-tests\fR
List available tests and exit.
.TP
\fB\-\-list\-plots\fR
List available plots for selected test and exit.
.TP
\fB\-V\fR, \fB\-\-version\fR
Show netperf\-wrapper version information and exit.
.TP
\fB\-h\fR, \fB\-\-help\fR
Show usage help message and exit.

.SH SUPPLIED TESTS
Test are supplied as Python files and can specify commands to run etc. For a
full list of the tests supported by netperf-wrapper, see the
\fB\-\-list\-tests\fR option.
.TP
\fBThe Realtime Response Under Load (RRUL) test\fR
.IP
This test exists in a couple of variants and is a partial implementation of the
RRUL specification as written by Dave Taht (see
\fIhttps://github.com/dtaht/deBloat/blob/master/spec/rrule.doc?raw=true\fR). It
works by running RTT measurement using ICMP ping and UDP roundtrip time
measurement, while loading up the link with eight TCP streams (four downloads,
four uploads). This quite reliably saturates the measured link (wherever the
bottleneck might be), and thus exposes bufferbloat when it is present.
.TP
\fBSimple TCP flow tests\fR
.IP
These tests combine a TCP flow (either in one direction, or both) with an ICMP
ping measurement. It's a simpler test than RRUL, but in some cases the single
TCP flow can be sufficient to saturate the link.
.TP
\fBUDP flood test\fR
.IP
This test runs \fIiperf\fR configured to emit 100Mbps of UDP packets targeted at
the test host, while measuring RTT using ICMP ping. It is useful for observing
latency in the face of a completely unresponsive packet stream.

.SH OUTPUT FORMATS
The following output formats are currently supported by netperf-wrapper:
.TP
\fBPlot output\fR (\fB\-f\fR \fIplot\fR)
.IP
Output test data as one of a series of graphical plots of timeseries data or
summarised as a CDF plot. Each test supplies a number of different plots; the
list of plots for a given test is output by the \fB\-\-list\-plots\fR switch
(which must be supplied along with a test name).
.IP
The plots are drawn by matplotlib, and can be displayed on the screen
interactively (requires a graphical display), or output to a file in svg, pdf,
ps and png formats. Using the \fB\-o\fR switch turns out file output (the file
format is inferred from the file name), while not supplying the switch turns on
the interactive plot viewer.

.TP
\fBTabulated output\fR (\fB\-f\fR \fIcsv\fR and \fB\-f\fR \fIorg_table\fR)
.IP
These formats output the numeric data in a tabulated format to be consumed by
other applications. The \fIcsv\fR output format is a comma-separated output that
can be imported into e.g. spreadsheets, while \fIorg_table\fR outputs a
tabulated output in the table format supported by Emacs org mode. The data is
output in text format to standard output, or written to a file if invoked with
the \fB\-o\fR parameter.

.TP
\fBStatistics output\fR (\fB\-f\fR \fIstats\fR)
.IP
This output format outputs various statistics about the test data, such as total
bandwidth consumed, and various statistical measures (min/max/mean/median/std
dev/variance) for each data source specified in the relevant test (this can
include some data sources not includes on plots). The data is output in text
format to standard output, or written to a file if invoked with the \fB\-o\fR
parameter.

.TP
\fBMetadata output\fR (\fB\-f\fR \fImetadata\fR)
.IP
This output format outputs the test metadata as pretty-printed json (also
suitable for human consumption). It is output as a list of objects, where each
object corresponds to the metadata of one test. Mostly useful for inspecting
metadata of stored data files.

.SH THE NETPERF-WRAPPER GUI
Netperf-wrapper comes equipped with a GUI to browse and plot previously captured
datasets. The GUI requires PyQt4; if this is installed, it can be launched with
the \fB--gui\fR parameter. Additionally, if netperf-wrapper is launched without
parameters and without a controlling terminal, the GUI will be launched
automatically.

The GUI can be used for interactively plotting previously captured datasets, and
makes it easy to compare results from several test runs. It presents a tabbed
interface to graphs of data files, allows dynamic configuration of plots, and
includes a metadata browser. For each loaded data file, additional data files
can be loaded and added to the plot, similar to what happens when specifying
multiple input files for plotting on the command line. A checkbox controls
whether the added data files are added as separate entries to the plot, or
whether they are used for scaling the output (mirroring the \fB--scale-mode\fR)
command line switch.

The GUI also incorporates matplotlib's interactive browsing toolbar, enabling
panning and zooming of the plot area,  dynamic configuration of plot and axis
parameters and labels and saving the plots to file. The exact dynamic features
supported depends on the installed version of matplotlib.

.SH RC FILE
Some of the command line options can be specified in an rc file. By default,
netperf-wrapper looks for this in \fI~/.netperf-wrapperrc\fR, but an alternative
location can be specified with the \fB--rcfile\fR command line option.

The rc file allows options to be specified globally, an optionally overridden
for specific tests. For an explanation of the options, refer to the annotated
example rc file, by default installed to
\fI/usr/share/doc/netperf-wrapper/netperf-wrapperrc.example\fR.

.SH BATCH FILES
Netperf-wrapper supports reading batch files to automate running several tests
and do setup/teardown of test environment etc. This greatly aids reproducibility
of tests.

The batch file format is based on the standard .ini file format, with sections
being split into three namespaces: Commands, starting with Command::, batches,
starting with Batch::, and arguments, starting with Arg::. Briefly, a batch is
the entity that will be run, commands can be run before or after each batch
iteration, and arguments allows parameterising batches.

\fBFIXME\fR: Expand this section; for now, try looking at the
\fIbatchfile.example\fR file supplied with the source code, and try to work
things out from there :).

.SH RUNNING TESTS WITH THE D-ITG TOOL
This version of netperf-wrapper has experimental support for running and parsing
the output of the D-ITG test tool (see
\fIhttp://traffic.comics.unina.it/software/ITG/\fR). Netperf-wrapper supports
parsing the one-way delay as measured by D-ITG. However, in order to do so, the
data needs to be collected at the receiver end, statistics extracted, and the
result passed back to netperf-wrapper on the sending side.

To perform this function, netperf-wrapper supports a control server which will
listen to XML-RPC requests, spawn an appropriate ITGRecv instance and, after the
test is done, parse its output and make it available for netperf-wrapper to
retrieve. This control server is available as a Python file that by default is
installed in \fI/usr/share/doc/netperf-wrapper/misc\fR. It currently requires a
patched version of D-ITG v2.8.1. The patch is also included in the same
directory.

Note that the D-ITG server is finicky and not designed with security in mind.
For this reason, the control server includes HMAC authentication to only allow
authenticated clients to run a test against the server; however there is
currently no support for enforcement of this in e.g. firewall rules. Please bear
this in mind when running a publicly reachable ITGRecv instance (with or without
the control server). Another security issue with the control server is that the
Python XML-RPC library by default is vulnerable to XML entity expansion attacks.
For this reason, it is highly recommended to install the 'defusedxml' library
(available at \fIhttps://pypi.python.org/pypi/defusedxml/\fR) on the host
running the control server. The server will try to find the library on startup
and refuse to run if it is not available, unless explicitly told otherwise.

.SH EXAMPLES
Run the \fIrrul\fR test against \fItestserver.example.com\fI:
.IP
netperf-wrapper rrul testserver.example.com
.PP
This produces no output, but saves the result in a datafile named after the
current date and time (in gzipped JSON format).
.PP
Show an interactive plot of a previously run test, which stored the data in
\fIdatafile.json.gz\fR (requires a working matplotlib and a graphical display):
.IP
netperf-wrapper -f plot datafile.json.gz
.PP
Combine multiple data files into one CDF plot:
.IP
netperf-wrapper -p icmp_cdf *.json.gz

.SH THE JSON DATA FORMAT
The aggregated test data is saved in a file called
\fI<test_name>-<date>.<title>.json.gz\fR (the title part is omitted if no title
is specified by the \fB-t\fR parameter). This file contains the data points generated
during the test, as well as some metadata. The top-level json object has three
keys in it: \fIx_values\fR, \fIresults\fR and \fImetadata\fR.
.PP
\fIx_values\fR is an array of the x values for the test data (typically the time
values for timeseries data).
.PP
\fIresults\fR is a json object containing the result data series. The keys are the
data series names; the value for each key is an array of y values for that data
series. The data array has the same length as the \fIx_values\fR array, but there
may be missing data points (signified by null values).
.PP
\fImetadata\fR is an object containing various data points about the test run. The
metadata values are read in as configuration parameters when the data set is
loaded in for further processing. Not all tests use all the parameters, but
they are saved anyway.
.PP
Currently the metadata values are:
.IP
\fINAME\fR: The test name.
.IP
\fITITLE\fR: Any extra title specified by the \fB-t\fR parameter when the test was run.
.IP
\fIHOSTS\fR: List of the server hostnames connected to during the test.
.IP
\fILOCAL_HOST\fR: The hostname of the machine that ran the test.
.IP
\fILENGTH\fR: Test length in seconds, as specified by the \fB-l\fR parameter.
.IP
\fITOTAL_LENGTH\fR: Actual data series length, after the test has added time to the \fILENGTH\fR.
.IP
\fISTEP_SIZE\fR: Time step size granularity.
.IP
\fITIME\fR: ISO timestamp of the time the test was initiated.
.IP
\fINOTE\fR: Arbitrary text as entered with the \fB--note\fR switch when the test was run.
.IP
\fINETPERF_WRAPPER_VERSION\fR: Version of netperf-wrapper that generated the
data file.
.IP
\fIIP_VERSION\fR: IP version used to run test (as specified by command line
parameters, or auto-detected from \fIgetaddrinfo()\fR if unspecified).
.PP
If the \fB\-\-extended\-metadata\fR switch is turned on, the following
additional values are collected and stored (to the extent they are available
from the platform):
.IP
\fIKERNEL_NAME\fR: The kernel name as reported by \fIuname -s\fR.
.IP
\fIKERNEL_RELEASE\fR: The kernel release as reported by \fIuname -r\fR.
.IP
\fIIP_ADDRS\fR: IP addresses assigned to the machine running netperf-wrapper.
.IP
\fIGATEWAYS\fR: IP addresses of all detected default gateways on the system,
and the interfaces they are reachable through. Only available if the
\fInetstat\fR binary is present on the system.
.IP
\fIEGRESS_INFO\fR: Egress interface, its qdisc, offload, driver and BQL configuration,
and (if applicable) the IP address of the next-hop router used to reach the test target.
The egress interface and next-hop router requires that the \fIip\fR binary is
present on Linux, but can be extracted from \fIroute\fR on BSD. Qdisc
information requires the \fItc\fR binary to be present, and offload information
requires \fIethtool\fR.
.PP
If the \fB--remote-metadata\fR is used, the extended metadata info is gathered
for each of the hostnames specified. This is gathered under the

CHANGES
2014-05-02: Add --subplot-combine option to combine multiple data sets into
            separate subplots rather than plotting them onto the same plot.

2014-05-02: Manage plot series colours to make them consistent when combining
            plots, and instead cycle line styles and markers to differentiate
            result sets in a combined plot.

2014-05-02: Fix plot scaling on box plots.

2014-05-02: Fix several bugs in the resizing of plots to make room for the
            legend (in interactive mode).

2014-04-30: Fix parsing of netperf output to not include spurious results at the end
            of UDP measurement flows giving a huge RTT because the measurement flow
            was lost during the test.

2014-04-30: Fix SIGINT behaviour so netperf-wrapper exits by SIGINT'ing itself after
            having cleaned up runners etc.

2014-04-30: Add batch run mode to automatically run several parameterised tests.

2014-04-09: On GUI startup, by default search for a running instance and pass
            the file name to that rather than starting a new instance.

2014-04-09: Add .desktop file and mime info for data files.

2014-04-09: Add toggle to invert y-axis on latency measurements.

2014-04-04: Add Q-Q plot type.

2014-03-24: Allow tests to specify axis labels explicitly. Use this to write
            something better than 'ms' on the latency axes.

2014-03-24: Add box plot type.

2014-03-23: Disable hostname lookup when loading data files or the GUI; should
            fix issue with hostname lookup failing when a datafile with an
            unreachable host is loaded.

2014-03-23: Add GUI for interactive browsing of results.

2014-03-20: rc-file should specify hosts rather than host.

2014-03-19: Record path towards machine running test in INGRESS_INFO of remote
            metadata object.

2014-03-19: Capture IP address info per-interface.

2014-03-16: Scale x axis to include all captures data points, even if they go beyond
            the configured test duration.

2014-03-16: Run ping with -n.

2014-03-05: Add remote metadata support.

2014-03-04: More extended metadata.

2014-03-01: Support an rcfile and supply an example file.

2014-03-01: Add metadata output formatter to extract metadata from stored data files.

2014-03-01: Add collection of several extended metadata items.

2014-01-22: Initial provisions for optionally gathering extended metadata on test runs.

2014-01-19: Add detection of supported netperf options and support passing -e to
            netperf for UDP_RR socket timeout (adds --socket-timeout parameter).

2014-01-18: Add rrul_be test using exclusively BE traffic.

2013-12-25: Don't crash on empty values in CDF plots.

2013-11-25: Release v0.6.1.

2013-11-25: Fix module rewrite in setup.py and numpy version output in --version.

2013-11-25: Release v0.6.0.

2013-10-31: Adjust spacing for title and annotation to work better with exported plots.

2013-09-27: Add description of output formats to man page.

2013-09-27: Add the --note argument to store arbitrary notes in the JSON metadata.

2013-09-27: Print custom title on separate line on plots.

2013-09-26: Add man page.

2013-09-25: Allow passing of data filenames and host names as unqualified arguments
            and attempt to guess what each argument is.

2013-09-25: Add matplotlib version to --version output.

2013-04-20: Manual removal of temporary files to get around issues on cygwin.

2013-04-18: Print Python version along with netperf version on -V.

2013-04-18: Fix python3.1 compatibility bug in gzip workaround.

2013-04-15: Release v0.5.6.

2013-04-15: Add rrul_noclassification test.

2013-04-15: Release v0.5.5.

2013-04-15: Add work-around for python2.6 version of GzipFile.

2013-04-15: Add option to disable automatic log-scaling of graphs.

2013-03-27: Add -z option to start y-axis at zero.

2013-03-27: Don't use suptitle() to draw annotation at the bottom of the plots (makes
            the annotation work in matplotlib 1.2.1).

2013-03-27: Release v0.5.3.

2013-03-27: Fix install bug in setup.py when creating packages.

2013-03-27: Release v0.5.2.

2013-03-26: Actually get the return code of a subprocess.

2013-03-26: Fix bug on aborting tests due to missing kill() stub for computing runners.

2013-03-26: Rework the way access checks are done for output files.

2013-03-18: Release v0.5.1.

2013-03-18: Compatibility fix for Python 3.1.

2013-03-18: Release v0.5.0.

2013-03-18: Do not crash if killing subprocess fails because it is already killed.

2013-03-18: Ensure compatibility with a wider range of python versions.

2013-03-18: Do not fail horribly on write permission errors.

2013-03-11: Preforking runner implementation (hopefully fixes issues on FreeBSD).

2013-03-11: Be smarter about picking a `ping` binary, trying first fping
            then falling back to straight ping.

2013-02-20: Don't crash on missing data, just print a warning.

2013-02-20: Workaround for bad exit codes on bsd.

2013-02-14: Release v0.4.2.

2013-01-25: Encode title in data set filename if set.

2013-01-22: Fix possible divide by zero error on certain data values.

2013-01-22: Get rid of the colons in the time stamps used for file names (they
            give rise to escaping headaches).

2012-12-20: Release v0.4.1.

2012-12-20: Fix cutoff logic when end cutoff value is 0.

2012-12-20: If there are missing data values when doing a CDF plot, the plot
            should cut off at a probability < 1; i.e. use the total size of the
            data (including null data points) as data set size when calculating
            probability.

2012-12-17: Release v0.4.0.

2012-12-15: Add stats formatter (-f stats) to output statistics on results data sets.

2012-12-13: Fix bug that caused smoothing to add new data points at the ends of
            series.

2012-12-13: Fix bug that caused primary axis to place ticks on the right hand
            side in dual-axis plots.

2012-12-13: Release v0.3.0.


netperf-wrapper.py
#!/usr/bin/env python2
## -*- coding: utf-8 -*-
##
## netperf-wrapper.py
##
## Author:   Toke Høiland-Jørgensen (toke@toke.dk)
## Date:     October 8th, 2012
## Copyright (c) 2012, Toke Høiland-Jørgensen
##
## This program is free software: you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation, either version 3 of the License, or
## (at your option) any later version.
##
## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with this program.  If not, see <http://www.gnu.org/licenses/>.

import sys

try:
    import optparse, os

    import aggregators, formatters, util
    from settings import settings, load



    config = util.DefaultConfigParser({'delay': 0})
    config.add_section('global')
    config.set('global', 'name', 'Netperf')
    config.set('global', 'iterations', '1')
    config.set('global', 'output', 'org_table')
    config.set('global', 'cmd_opts', '-P 0 -v 0')
    config.set('global', 'cmd_binary', '/usr/bin/netperf')

    if __name__ == "__main__":
        try:
            settings,results = load()

            aggregator_name = settings.AGGREGATOR
            classname = util.classname(aggregator_name, "Aggregator")
            if hasattr(aggregators, classname):
                agg = getattr(aggregators, classname)()
            else:
                raise RuntimeError("Aggregator not found: '%s'" % aggregator_name)

            for s in settings.DATA_SETS.items():
                agg.add_instance(*s)

            formatter_name = util.classname(settings.FORMAT, 'Formatter')
            if hasattr(formatters, formatter_name):
                formatter = getattr(formatters, formatter_name)(settings.OUTPUT)
            else:
                raise RuntimeError("Formatter not found.")

            if not results:
                results = agg.postprocess(agg.aggregate(results))
                results.dump_dir(os.path.dirname(settings.OUTPUT) or ".")
            formatter.format(results)

        except RuntimeError, e:
            sys.stderr.write(u"Error occurred: %s\n"% unicode(e))
            sys.exit(1)
except KeyboardInterrupt:
    try:
        agg.kill_runners()
    except NameError:
        pass
    sys.exit(1)
